src/ir/module-utils.h

```diff
@@ -448,10 +449,10 @@ collectSignatures(Module& wasm,
   for (auto& curr : wasm.functions) {
     counts[curr->sig]++;
     for (auto type : curr->vars) {
-      if (type.isRef()) {
-        auto heapType = type.getHeapType();
-        if (heapType.isSignature()) {
-          counts[heapType.getSignature()]++;
+      counts.maybeNote(type);
+      if (type.isTuple()) {
+        for (auto t : type) {
+          counts.maybeNote(t);
         }
       }
```

Comment:
```suggestion
```
Because we are doing tuple handling below, I guess we don't need to do this here again?

Reply:
I don't think that's right (but I could be confused again).

If we have a tuple local that contains two different TFRs, `maybeNote` on the local type won't do anything because `maybeNote` only increments counts if the type is a TFR, not if it is a tuple. But we still want to count those two TFRs in the tuple because in the output binary they will be separate locals that refer to their type indices. (Although aren't local types run length encoded or something? I don't think we've been accounting for that, but I think that's fine for now because that would be complicated. Overcounting is better than undercounting here.)

Reply:
From the line 470+, we have
```cpp
    for (Type type : {sig.params, sig.results}) {
      for (auto element : type) {
        ...
```
Doesn't the `for (auto element : type)` take care of tuples?

Reply:
Oh no nevermind, locals may not be itself a signature but contain tuples I guess...

Reply:
Yes, this line handles a local that is a tuple. So `(i32 f32 f64)`, something like that. We just look at the elements here for any signatures, and note those. Tuples aren't recursive so we can stop looking for tuples at that point.

The loop `for (auto element : type)` lower down handles multiple return values and multiple parameters, in a signature (which can be recursive). I actually don't know if each of those is an actual tuple type - would `isTuple` return true there..? If it would, then maybe this can all be simplified somehow...

Reply:
The params and results of a signature may or may not be tuples. For example, if a signature has no params, then the params would be `Type::none`, which is not a tuple. Similarly, if there is just one result, the results would be that type rather than a tuple.